File: projects/order-creation-queries/app/src/main/java/com/robinfood/app/usecases/getorderhistory/GetOrderHistoryUseCase.java
package com.robinfood.app.usecases.getorderhistory;

import com.robinfood.app.usecases.gettokenbusinesscapability.IGetTokenBusinessCapabilityUseCase;
import com.robinfood.core.dtos.EntityDTO;
import com.robinfood.core.dtos.orderhistory.request.OrderHistoryRequestDTO;
import com.robinfood.core.dtos.orderhistory.response.OrderHistoryItemDTO;
import com.robinfood.core.enums.Result;
import com.robinfood.repository.orderhistory.IOrderHistoryRepository;
import lombok.extern.slf4j.Slf4j;

/**
 * Implementation of IGetOrderHistoryUseCase
 */
@Slf4j
public class GetOrderHistoryUseCase implements IGetOrderHistoryUseCase {

    private final IOrderHistoryRepository orderHistoryRepository;
    private final IGetTokenBusinessCapabilityUseCase getTokenBusinessCapabilityUseCase;

    public GetOrderHistoryUseCase(
            IOrderHistoryRepository orderHistoryRepository,
            IGetTokenBusinessCapabilityUseCase getTokenBusinessCapabilityUseCase
    ) {
        this.orderHistoryRepository = orderHistoryRepository;
        this.getTokenBusinessCapabilityUseCase = getTokenBusinessCapabilityUseCase;
    }

    @Override
    public Result<EntityDTO<OrderHistoryItemDTO>> invoke(OrderHistoryRequestDTO orderHistoryRequestDTO) {

        var token = getTokenBusinessCapabilityUseCase.invoke();

        return orderHistoryRepository.getOrderHistory(
                orderHistoryRequestDTO,
                token.getAccessToken()
        );
    }
}

File: projects/posv2/app/src/main/kotlin/com/robinfood/app/usecases/getorderhistory/GetOrderHistoryUseCase.kt
package com.robinfood.app.usecases.getorderhistory

import com.robinfood.core.dtos.HistoryDTO
import com.robinfood.core.dtos.orderhistory.OrderHistoryRequestDTO
import com.robinfood.core.enums.Result
import com.robinfood.repository.orderhistory.IOrderHistoryRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * Implementation of IGetOrderHistoryUseCase
 */
class GetOrderHistoryUseCase(
        private val orderHistoryRepository: IOrderHistoryRepository
) : IGetOrderHistoryUseCase {

    private val log: Logger = LoggerFactory.getLogger(this.javaClass)

    override suspend fun invoke(
            token: String,
            orderHistoryRequestDTO: OrderHistoryRequestDTO
    ): Result<HistoryDTO> {

        log.info("Init Get Order History Use Case");

        return orderHistoryRepository.getOrderHistory(
                token,
                orderHistoryRequestDTO
        )
    }
}